Add documentation of the state of acceleration in Axom (#620)
LLNL/axom: src/axom/core/docs/sphinx/core_acceleration.rst

.. ## Copyright (c) 2017-2021, Lawrence Livermore National Security, LLC and
.. ## other Axom Project Developers. See the top-level LICENSE file for details.
.. ##
.. ## SPDX-License-Identifier: (BSD-3-Clause)

******************************************************
Core acceleration
******************************************************

> 

Reviewer: We should probably define "accelerators" here to provide the reader with some context.
I.e. this page is about support for different programming model backends, like OpenMP and CUDA.

This page might be helpful: https://raja.readthedocs.io/en/main/feature/policies.html

Reviewer 2: I think the link is now https://raja.readthedocs.io/en/main/sphinx/user_guide/feature/policies.html.

I added some high-level discussion.

Axom Core provides support to ease the implementation of user applications that

> intend to support execution utilizing accelerators. In order to supply this,

Reviewer 2: Updated using part of your suggestion

Axom provides a simple API to access memory and computation, using RAJA

> and Umpire to access accelerators.

Reviewer: Please add links to RAJA and Umpire.

Reviewer 2: added

.. warning:: 
   You do need to compile with RAJA and Umpire to fully leverage this Axom functionality. 

>    While both submodules will operate without these tools, they will offer little.

Reviewer 3: It's not clear to me what you mean by "submodules" here.

Reviewer 2: rephrased

>    You do need to compile with RAJA and Umpire to fully leverage this Axom functionality. 
>    While both memory management and execution spaces in Axom will operate without these tools, they will offer little.

Reviewer: Suggestion: The following might be a clearer way to express this point:
```suggestion
   Axom's memory management and execution space APIs have default implementations when Axom is not
   configured with Umpire and RAJA, respectively, and can be used in all Axom configurations.
```

Reviewer 2: changed (thanks)

> The memory management submodule offers a singular interface which allows the user to leverage either C++ memory functions or Umpire, 

Reviewer: ```suggestion
The memory management API allows the user to leverage either C++ memory functions or Umpire, 
```

Reviewer: Suggestion: Adding subsections for memory management and execution spaces might make it easier for readers to get a quick feel for the page.

Reviewer 2: I lightly rephrased this part.

Reviewer 2: I think this documentation can always be improved by further restructuring or editing, but will serve its purpose as-is.

depending on the availability of Umpire at compilation. It supports a simple operation set: allocate, deallocate, reallocate, and copy. 

> If Umpire is in use, an allocator can be specified -- see Umpire documentation(make this a link) for more details. Note that the fall-back to

Reviewer 3: Here's the link to add: https://readthedocs.org/projects/umpire/

Reviewer 2: added

C++ memory functions is automatic, so one piece of code can handle standard C++ or C++ with Umpire. However, to use advanced features, such as 

> accessing unified memory, Umpire must be enabled, otherwise errors will occur at compilation.

Reviewer: It seems that we're making this point a few times. Would it be possible to streamline this a bit?

Here is an example of using Axom’s memory management tools:

.. literalinclude:: ../../examples/core_acceleration.cpp
   :start-after: _membasic_start
   :end-before: _membasic_end
   :language: C++

Throughout Axom, acceleration is increasingly supported. Both internally, and to support users, Axom Core offers an 

> interface that, using RAJA and Umpire internally, provides easy access to for-loop level acceleration via the for-all model, 

Reviewer 3: better stated if you replace "for-all model" with "parallel-for idiom"

Reviewer 2: replaced

which applies a given lambda function for every index in range.  

> Here is an example of Axom in motion with basic sequential execution:

Reviewer 3: Reword to "Here is an Axom example showing sequential execution"

Reviewer 2: reworded

.. literalinclude:: ../../examples/core_acceleration.cpp
   :start-after: _exebasic_start
   :end-before: _exebasic_end
   :language: C++

Here's the same loop from the above snippet, this time with CUDA:

.. literalinclude:: ../../examples/core_acceleration.cpp
   :start-after: _cudaexebasic_start
   :end-before: _cudaexebasic_end
   :language: C++

For more advanced functionality, see the `RAJA documentation <https://raja.readthedocs.io>`_, and the `Umpire documentation <https://umpire.readthedocs.io>`_. Axom 

> allows for these to be used in lieu of Axom’s API.  

Reviewer: Consider rephrasing the following:
> Axom allows for these to be used in lieu of Axom’s API.  

I know what this means, but general readers might not.

I think the point is that one can directly use RAJA, and does not need to use the `axom::for_all` API ?

Reviewer 2: Rephrased.

.. ## Copyright (c) 2017-2021, Lawrence Livermore National Security, LLC and
.. ## other Axom Project Developers. See the top-level LICENSE file for details.
.. ##
.. ## SPDX-License-Identifier: (BSD-3-Clause)

******************************************************
Core acceleration
******************************************************

Axom's `core` component provides several utilities for user applications that intend
to support execution on hardware accelerators.
Axom lets users control execution space using RAJA and memory space using Umpire.
As noted in the `RAJA documentation <https://raja.readthedocs.io/en/main/index.html>`_,
developers of high-performance computing applications have many options for running
code: on the CPU, using OpenMP, or on GPU hardware accelerators, and the options are
constantly developing.  RAJA controls *where* code runs; Umpire moves data between
memory spaces.

> .. warning::

Reviewer: Minor: This feels like a `note` rather than a warning.

```suggestion
.. note::
```
